src/tools/ceph_immutable_object_cache/ObjectCacheStore.cc

```diff
@@ -0,0 +1,273 @@
+    delete read_buf;
+    return ret;
+  }
+
+  if (ret == -ENOENT) {
+    // object is empty
+    ret = 0;
+  }
+
+  if (ret < object_size) {
+    // object is partial, fill with '0'
+    read_buf->append(std::string(object_size - ret, '0'));
+  }
+
+  if (dir_num > 0) {
+    auto const pos = cache_file_name.find_last_of('.');
+    cache_dir = cache_dir + "/" + std::to_string(stoul(cache_file_name.substr(pos+1)) % dir_num);
+  }
+  // write to cache
+  SyncFile cache_file(m_cct, cache_dir + "/" + cache_file_name);
+  cache_file.create();
+
+  ret = cache_file.write_object_to_file(*read_buf, object_size);
+  if (ret < 0) {
+    lderr(m_cct) << "fail to write cache file" << dendl;
+
+    m_policy->update_status(cache_file_name, OBJ_CACHE_NONE);
+    delete read_buf;
+    return ret;
+  }
+
+  // update metadata
+  assert(OBJ_CACHE_SKIP == m_policy->get_status(cache_file_name));
+  m_policy->update_status(cache_file_name, OBJ_CACHE_PROMOTED);
+  assert(OBJ_CACHE_PROMOTED == m_policy->get_status(cache_file_name));
+
+  delete read_buf;
+  return ret;
+
+  evict_objects();
+}
+
+int ObjectCacheStore::lookup_object(std::string pool_name,
+    std::string vol_name, std::string object_name) {
+  ldout(m_cct, 20) << "object name = " << object_name
+                   << " in pool: " << pool_name << dendl;
+
+  int pret = -1;
+  cache_status_t ret = m_policy->lookup_object(pool_name + object_name);
+
+  switch(ret) {
+    case OBJ_CACHE_NONE: {
+      pret = do_promote(pool_name, vol_name, object_name);
+      if (pret < 0) {
+        lderr(m_cct) << "fail to start promote" << dendl;
+      }
+      return -1;
+    }
+    case OBJ_CACHE_PROMOTED:
+      return 0;
+    case OBJ_CACHE_SKIP:
+      return -1;
+    default:
+      lderr(m_cct) << "unrecognized object cache status." << dendl;
+      assert(0);
+  }
+}
+
+int ObjectCacheStore::promote_object(librados::IoCtx* ioctx, std::string object_name,
+                                     librados::bufferlist* read_buf, uint64_t read_len,
+                                     Context* on_finish) {
+  ldout(m_cct, 20) << "object name = " << object_name
+                   << " read len = " << read_len << dendl;
+
+  auto ctx = new FunctionContext([on_finish](int ret) {
+    on_finish->complete(ret);
+  });
+
+  librados::AioCompletion* read_completion = librbd::util::create_rados_callback(ctx);
+  int ret = ioctx->aio_read(object_name, read_completion, read_buf, read_len, 0);
+  if(ret < 0) {
+    lderr(m_cct) << "fail to read from rados" << dendl;
+  }
+  read_completion->release();
+
+  return ret;
+}
+
+int ObjectCacheStore::evict_objects() {
+  ldout(m_cct, 20) << dendl;
+
+  std::list<std::string> obj_list;
+  m_policy->get_evict_list(&obj_list);
+  for (auto& obj: obj_list) {
+    do_evict(obj);
+  }
+}
+
+int ObjectCacheStore::do_evict(std::string cache_file) {
+  ldout(m_cct, 20) << "file = " << cache_file << dendl;
+
+  //TODO(): need a better way to get file path
```

Comment:
For better efficiency when reading from lots of small 4M objects, I did a simple hashing to distribute those objects to 10 different folders. otherwise the file system overhead would be high. The layout of a caching folder is:
(poolname_rbdid/shard_id/filename)
rbd_31686b8b4567/0/rbdrbd_data.31686b8b4567.0000000000000daa
rbd_31686b8b4567/1/...
rbd_31686b8b4567/2/...
...
rbd_31686b8b4567/9/...

Currently in the metastore, only the file name(like rbdrbd_data.31686b8b4567.0000000000000daa) is recorded.  The file path could be conducted with the file name so we could remove the cache file.

Reply:
OK -- can you remove all the RBD-unique references to keep it generic?

Reply:
done. all rbd related logic are removed. the daemon can do promotion/eviction based on poolname + objectid